### ISSUE TITLE
RFC: L3 Nightly Regression for Qwen-Image + FlowGRPO

### ISSUE THREAD

**Issue by wtomin** [verl-project/verl-omni#275]
### Feature request

This RFC describes the architecture for **nightly regression**: run 20 steps of FlowGRPO training on local tiny random `Qwen-Image` weights with a single fixed sample, use a local tiny random Qwen-VL reward model, and detect **numerical drift** and **performance regressions** in one job.

The core idea is to complete both checks in a single run while avoiding cross-interference via **step partitioning**:

- **Steps 1-2**: dump forward intermediates and LoRA gradients; compare against baseline at tensor granularity.
- **Steps 3-20**: train normally and record timing / throughput / memory; compare performance against baseline (dump steps are excluded from performance stats due to I/O overhead).

All diagnostic logic lives **under `tests/nightly/`** and does not touch `verl_omni/` production code.

---

## Overall Architecture

```text
┌─────────────────────────────────────────────────────────────────┐
│  tests/nightly/qwen_image_flowgrpo_single_sample/               │
│  ┌──────────────┐  ┌──────────────────┐                         │
│  │ run.py       │→ │ install_debug_   │                         │
│  │ (entrypoint) │  │ hooks.py         │                         │
│  └──────┬───────┘  │ (hooks + I/O)    │                         │
│         │          └──────────────────┘                         │
│         │ env: DEBUG_DUMP_*                                     │
│         ▼                                                       │
│  ┌──────────────────────────────────────────────────────────┐   │
│  │ verl_omni.trainer.main_diffusion  (unchanged)            │   │
│  └──────────────────────────────────────────────────────────┘   │
│         │ post-run                                              │
│         ▼                                                       │
│  ┌─────────────────┐  ┌──────────────────┐                      │
│  │ compare_dumps.py│  │ collect_metrics.py│                     │
│  │ (dump steps)    │  │ (skip dump steps) │                     │
│  └─────────────────┘  └──────────────────┘                      │
└─────────────────────────────────────────────────────────────────┘
```

Lifecycle of one nightly run:

```mermaid
flowchart LR
    A[prepare single repeated data] --> B[run.py]
    B --> C[20-step training]
    C --> D{step}
    D -->|in DEBUG_DUMP_STEPS| E[dump tensors]
    D -->|all steps| F[log metrics]
    E --> G[compare_dumps]
    F --> H[collect_metrics]
    G --> I[pass/fail]
    H --> I
```

---

### Fixed Workload

| Dimension | Choice | Rationale |
| --- | --- | --- |
| Model | Local tiny random Qwen-Image, default `~/models/tiny-random/Qwen-Image` | Exercises the Qwen-Image FlowGRPO path without using full production weights |
| Algorithm | FlowGRPO + LoRA | Primary RL training path |
| Data | 1 text-to-image sample, deterministically repeated as needed | Eliminates sampling randomness while satisfying batch/group shape requirements |
| Rollout | vllm-omni, n=4 | Ensures nonzero reward std within the GRPO group |
| Training | FSDP2, 4 GPUs, 20 steps | Aligns with typical nightly machine configs |
| Reward | Tiny random Qwen-VL reward model, default `~/models/tiny-random/qwen3-vl` | Exercises the visual reward-model path with local tiny random weights |

### Randomness Control

Fix seeds, disable shuffle, disable random SDE window selection, use a single rollout worker, and set `enforce_eager` to keep run-to-run noise within acceptable bounds. Precision comparison uses **relative-error thresholds** (not bit-exact). See plan §4 for details.

### Three Dump Points

```text
Driver (PolicyGradientRayTrainer)
  └─ after compute_advantage, before _update_actor
       └─ driver_forward/   responses, logprobs, latents, advantages, ...

Worker (DiffusersFSDPEngine)
  ├─ after postprocess_batch_func
  │    └─ actor_forward/    log_probs, loss
  └─ before clip_grad_norm (optimizer_step)
       └─ lora_gradients/   LoRA param grads (local shard)
```

### Hook Design

Use the smallest hook points that match the current dataflow.

| Process | Hook target | Purpose | Rationale |
| --- | --- | --- | --- |
| Driver | `PolicyGradientRayTrainer._update_actor` | Dump post-advantage actor input and inject debug metadata into the TensorDict sent to workers. | In the current loop, `_update_actor(batch)` is called immediately after `compute_advantage(...)`, so its input is the right driver-side snapshot. This is less brittle than wrapping the entire `fit()` loop. |
| Worker | `DiffusersFSDPEngine._run_forward_backward_batch` | Dump actor forward outputs and loss after `postprocess_batch_func(...)`. | This gives tensors after micro-batch and diffusion-step aggregation, matching what the engine returns before `model_output` is dropped by the worker. |
| Worker | `DiffusersFSDPEngine.forward_step` or a narrow wrapper inside `_run_forward_backward_batch` | Optionally dump per-diffusion-step signatures when aggregate output is insufficient. | Per-step drift is easier to localize than only comparing fully stacked outputs. Keep this optional to limit I/O. |
| Worker | `DiffusersFSDPEngine.optimizer_step` | Dump LoRA gradients before clipping and optimizer update. | This captures the trainable gradient state before it is mutated by clipping or stepping. |

Hook installation must be idempotent. Each wrapper should check a marker such as `__nightly_debug_wrapped__` or `__wrapped__` before replacing a method.

### Step Identity

Every dump must include a structured identity, not only `step_{n}`:

```json
{
  "global_step": 1,
  "ppo_epoch": 0,
  "mini_batch_idx": 0,
  "micro_batch_idx": 0,
  "diffusion_step": 3,
  "rank": 0,
  "world_size": 4,
  "dp_rank": 0,
  "dp_size": 4
}
```

The exact fields can vary by hook point, but `global_step`, `rank`, and `world_size` are required. Worker-side dumps should also record mini-batch and diffusion-step identity when available.

This matters because `TrainingWorker.train_mini_batch(...)` re-splits the actor batch and uses rank-dependent dataloader seeds. A path like `actor_forward/step_1/rank_0/` is not precise enough to explain which mini-batch or diffusion timestep produced a tensor.

Baseline layout (relative to repo root):

```text
outputs/debug_dumps/
├── current/     # written each run (DEBUG_DUMP_DIR)
└── baseline/    # golden copy: dump dirs + metrics.json
```

---

### Dual-Track Validation

| Track | Script | Input | Comparison range |
| --- | --- | --- | --- |
| Precision | `compare_dumps.py` | `baseline/` vs `current/` dump dirs | steps in `DEBUG_DUMP_STEPS` |
| Performance | `collect_metrics.py` | console log + `memory.log` | steps `perf_skip_steps+1` .. total |

Precision comparison checks `max_rel_err`, `max_abs_err`, and `cos_sim` per tensor. Performance comparison checks timing / throughput / memory deviation from baseline (initial threshold ~5%).

The performance JSON keeps both `metrics_all_steps` (full sequence) and `metrics_for_compare` (subset after skipping dump steps) for post-hoc debugging. The derived `perf_skip_steps` is written into the JSON for auditability.

### Baseline Lifecycle

1. **Bootstrap**: first green run does not fail; store dump-step artifacts and performance JSON as baseline.
2. **Nightly**: each run compares against baseline; alert if either track fails.
3. **Update**: baselines may be updated only from scheduled green runs, with commit / CUDA / GPU model recorded.

LoRA gradients are FSDP shards; baseline comparison requires the **same GPU count, FSDP topology, and launch configuration**.

---

## Directory Layout

All implementation lives under:

```text
tests/nightly/qwen_image_flowgrpo_single_sample/
├── create_single_sample_data.py
├── run.py                          # entrypoint
├── run_qwen_image_flowgrpo_single_sample.sh
├── install_debug_hooks.py
├── compare_dumps.py
├── collect_metrics.py
└── README.md
```

---

## Follow-Up Work

1. Implement all scripts under the nightly directory.
2. First baseline run + 2-3 repeats on the same commit to calibrate rtol and performance noise.
3. Wire into a scheduled CI workflow.
4. If the pattern works, generalize as a template for other model nightlies (copy hook framework to a new directory, swap Hydra overrides).
